File: src/main/java/com/github/sorusclient/client/ui/framework/Component.java
package com.github.sorusclient.client.ui.framework;

import com.github.sorusclient.client.adapter.event.KeyEvent;
import com.github.sorusclient.client.adapter.event.MouseEvent;
import com.github.sorusclient.client.ui.framework.constraint.Constraint;
import com.github.sorusclient.client.ui.framework.constraint.Flexible;
import com.github.sorusclient.client.ui.framework.constraint.Side;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public abstract class Component {

    protected Constraint x = new Side(Side.ZERO), y = new Side(Side.ZERO);
    protected Constraint width = new Flexible(), height = new Flexible();

    protected Runtime runtime;

    protected Container parent;

    private final List<String> storedState = new ArrayList<>();
    private final Map<String, Consumer<Map<String, Object>>> onStateUpdates = new HashMap<>();

    public Component() {
        this.addStoredState("selected");
        this.addStoredState("hasInit");
    }

    public Component setX(Constraint x) {
        this.x = x;
        return this;
    }

    public Component setY(Constraint y) {
        this.y = y;
        return this;
    }

    public Component setWidth(Constraint width) {
        this.width = width;
        return this;
    }

    public Component setHeight(Constraint height) {
        this.height = height;
        return this;
    }

    public Component addStoredState(String storedState) {
        this.storedState.add(storedState);
        return this;
    }

    public Component addOnStateUpdate(String state, Consumer<Map<String, Object>> onStateUpdate) {
        this.onStateUpdates.put(state, onStateUpdate);
        return this;
    }

    public Runtime getRuntime() {
        return runtime;
    }

    public void setParent(Container parent) {
        this.parent = parent;
    }

    public abstract class Runtime {

        public final Map<String, Object> state = new HashMap<>();

        public abstract void render(double x, double y, double width, double height);

        public abstract Container.Runtime getParent();

        public abstract double[] getCalculatedPosition();

        public abstract void setX(double x);
        public abstract void setY(double y);

        public abstract double getX();
        public abstract double getY();

        public abstract double getWidth();
        public abstract double getHeight();

        public abstract double getPadding();

        public Runtime() {
            this.setState("hasInit", false);
        }

        public void setState(String id, Object value) {
            if (Component.this.storedState.contains(id)) {
                this.state.put(id, value);
                this.onStateUpdate(id, value);
            } else {
                this.getParent().setState(id, value);
            }
        }

        public void onStateUpdate(String id, Object value) {
            Consumer<Map<String, Object>> onStateUpdate = Component.this.onStateUpdates.get(id);
            if (onStateUpdate != null) {
                Map<String, Object> state = this.getAvailableState();
                onStateUpdate.accept(state);
                this.setAvailableState(state);
            }
        }

        public Object getState(String id) {
            if (Component.this.storedState.contains(id)) {
                return this.state.get(id);
            } else {
                if (Component.this.parent != null) {
                    return this.getParent().getState(id);
                } else {
                    return null;
                }
            }
        }

        public Map<String, Object> getAvailableState() {
            Map<String, Object> availableState = new HashMap<>();

            if (Component.this.parent != null) {
                availableState.putAll(this.getParent().getAvailableState());
            }

            availableState.putAll(this.state);

            return availableState;
        }

        public void setAvailableState(Map<String, Object> state) {
            for (Map.Entry<String, Object> entry : state.entrySet()) {
                if (!entry.getValue().equals(this.getState(entry.getKey()))) {
                    this.setState(entry.getKey(), entry.getValue());
                }
            }
        }

        public void setHasInit(boolean hasInit) {
            this.setState("hasInit", hasInit);
        }

        public abstract boolean handleMouseEvent(MouseEvent event);
        public abstract void handleKeyEvent(KeyEvent event);

    }

}

File: src/main/kotlin/com/github/sorusclient/client/ui/framework/Component.kt
package com.github.sorusclient.client.ui.framework

import com.github.sorusclient.client.adapter.event.KeyCharEvent
import com.github.sorusclient.client.adapter.event.KeyEvent
import com.github.sorusclient.client.adapter.event.MouseEvent
import com.github.sorusclient.client.ui.framework.constraint.Constraint
import com.github.sorusclient.client.ui.framework.constraint.Flexible
import com.github.sorusclient.client.ui.framework.constraint.Side
import java.util.function.Consumer

abstract class Component {
    var x: Constraint = Side(Side.ZERO)
    var y: Constraint = Side(Side.ZERO)
    var width: Constraint = Flexible()
    var height: Constraint = Flexible()
    lateinit var runtime: Runtime
    var parent: Container? = null
    val storedState: MutableList<String> = ArrayList()
    val onStateUpdate: MutableMap<String, (MutableMap<String, Any>) -> Unit> = HashMap()

    init {
        storedState += "selected"
        storedState += "hasInit"
        storedState += "hidden"
        storedState += "clicked"
        storedState += "hovered"
        storedState += "interacted"
    }

    abstract inner class Runtime {
        val state: MutableMap<String, Any> = HashMap()
        abstract fun render(x: Double, y: Double, width: Double, height: Double)
        abstract fun getParent(): Container.Runtime?
        abstract val calculatedPosition: DoubleArray

        var x: Double = 0.0
        var y: Double = 0.0

        abstract val leftPadding: Double
        abstract val rightPadding: Double
        abstract val topPadding: Double
        abstract val bottomPadding: Double

        abstract val width: Double
        abstract val height: Double

        init {
            setState("hasInit", false)
            setState("hidden", false)
            setState("clicked", false)
            setState("hovered", false)
            setState("selected", false)
            setState("interacted", false)
        }

        abstract fun onInit()

        fun setState(id: String, value: Any) {
            if (storedState.contains(id)) {
                state[id] = value
                onStateUpdate(id, value)
            } else {
                getParent()!!.setState(id, value)
            }
        }

        open fun onStateUpdate(id: String, value: Any) {
            val onStateUpdate = onStateUpdate[id]
            if (onStateUpdate != null) {
                val state = availableState
                onStateUpdate(state)
                availableState = state
            }
        }

        fun getState(id: String?): Any? {
            return if (storedState.contains(id)) {
                state[id]
            } else {
                if (parent != null) {
                    getParent()!!.getState(id)
                } else {
                    null
                }
            }
        }

        var availableState: MutableMap<String, Any>
            get() {
                val availableState: MutableMap<String, Any> = HashMap()
                if (parent != null) {
                    availableState.putAll(parent!!.runtime.availableState)
                }
                availableState.putAll(state)
                return availableState
            }
            set(state) {
                for ((key, value) in state) {
                    if (value != getState(key)) {
                        setState(key, value)
                    }
                }
            }

        open fun setHasInit(hasInit: Boolean) {
            setState("hasInit", hasInit)
        }

        open fun onClose() {

        }

        abstract fun handleMouseEvent(event: MouseEvent): Boolean
        abstract fun handleKeyEvent(event: KeyEvent): Boolean
        abstract fun handleKeyCharEvent(event: KeyCharEvent): Boolean
    }
}